sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -5207,6 +5207,16 @@ object SQLConf {
     .booleanConf
     .createWithDefault(false)
 
+  val ORDERING_AWARE_LIMIT_OFFSET = buildConf("spark.sql.orderingAwareLimitOffset")
+    .internal()
+    .doc("When set to true, a local sort will be inserted between GlobalLimitExec and " +
+      "single-partition ShuffleExchangeExec, if the underlying plan produces sorted data. " +
+      "This is because shuffle reader in Spark fetches shuffle blocks in a random order and " +
+      "can not preserve the data ordering, while LIMIT/OFFSET must preserve ordering.")
+    .version("4.0.0")
+    .booleanConf
+    .createWithDefault(true)
```

Comment:
Since this is `true` by default, we need a migration guide, don't we, @cloud-fan ?

Reply:
This is a correctness bug with perf penalty, shall we just mention that LIMIT/OFFSET query may get slower after Spark 4.0?

Reply:
I see this as more of an "escape hatch" flag in the unlikely scenario that a user had a workload which was previously using an incorrect execution plan but for which they are okay with re-enabling incorrect behavior for performance reasons. 

Given that the originally problematic plan is not used in some of the most common SQL scenarios (where, typically, either TakeOrderedAndProjectExec or CollectLimitExec are used), I think we might hit some trickiness in how we message this in a guide in a way that will be comprehensible to end users: we probably don't want to give overly-vague guidance which could be misinterpreted as "if you have <generic sounding performance problem> flip this setting", but a more narrow and precise description might be similarly unclear to users.

I'm not entirely sure what the right move is here, but I'm provisionally okay with tackling the exact wording in a documentation followup if needed.

Reply:
Given the circumstance, I added `correctness` label to the JIRA issue.